### ISSUE TITLE
[question] any instruction on train multi custom classes?

### ISSUE THREAD

**Issue by eisneim** [MarvinTeichmann/KittiBox#13]
Hi, suppose i  wanna use kittiBox to detect road-sign, billboard, monitor screen and totally 7 types of object, i seem like i have to modify AnnotationLib.py to make it parse annotation rect with class number.

thanks!

**Comment by MarvinTeichmann** [comment]
Hi,

yes, you do. In addition you have to modify the loss. Currently it does this:

> true_classes = tf.reshape(tf.cast(tf.greater(classes, 0), 'int64'), [outer_ size * hypes['rnn_len']])

This is implementation is actually quite ugly and only works for two classes. Make sure that you generate a vector representing class labels as one-hot vector or as numerical class vector. You can then use this vector in combination with: `tf.nn.sparse_softmax_cross_entropy_with_logits` or `tf.nn.softmax_cross_entropy_with_logits` depending on your vector encoding.

Btw. I am also going to try to reduce the annonlib dependency. So you might want to keep an eye on the next one or two commits.

**Comment by eisneim** [comment]
@MarvinTeichmann great! I'll give it a try

**Comment by MarvinTeichmann** [comment]
As mentioned in the commit messages, keep a close look at the modifications of  	`fab6bae` and `bb25fc6`. The functions in `fab6bae` can help you to visualize what is going on during data loading. `bb25fc6` removes most of the dependency to AnnoLib. Data loading is now done directly fromt he kitti files. You might want to do somethink along the lines of `bb25fc6` for you stuff.

**Comment by eisneim** [comment]
@MarvinTeichmann wow, thank you so much

**Comment by eisneim** [comment]
@MarvinTeichmann after spending few hours reading the code, i just got some basic understanding of it, and there is an error initializing tf computation graph:
```
File "/Users/eisneim/www/deepLearning/KittiBox/hypes/../decoder/fastBox.py", line 382, in loss
    name='reg_loss')
File "/Library/Python/2.7/site-packages/tensorflow/python/ops/math_ops.py", line 1951, in add_n
    raise ValueError("inputs must be a list of at least one Tensor with the "
```
since YOLO is working OK for my project, so i guess i'll try to fine-tune YOLO model first and keeping an eye on KittiBox, definitely gonna test out KittiBox later
Thanks

**Comment by MarvinTeichmann** [comment]
Did you run `git submodule update` after your last pull? The reason for the error is, that you are using an old tensorflow-fcn version together with the new kittibox version. (Or vice versa).

**Comment by eisneim** [comment]
oops! that's my mistake, issue fixed

**Comment by MarvinTeichmann** [comment]
@eisneim may I asked which Yolo implementation you are using? I would like to add another baseline to the paper.

**Comment by eisneim** [comment]
@MarvinTeichmann  [darknet](https://github.com/thtrieu/darkflow) for training (using YOLOv2 network) use tensorflow implementation [darkflow](https://github.com/thtrieu/darkflow) for other stuff

**Comment by yangyuekai** [comment]
Hi, I want to know that did you change the code to multi class? If you did,can you give me a instruction? I have no idea on the change of code. Thanks a lot if you can answer me. It's better if you can help me.
